### PR TITLE
Add integration for MLflow AI Gateway

### PR DESCRIPTION
**Pull request by harupy**
- Adds integration for MLflow AI Gateway (this will be shipped in MLflow 2.5 this week).

Manual testing:

```sh
# Move to mlflow repo
cd /path/to/mlflow

# install langchain
pip install git+https://github.com/harupy/langchain.git@gateway-integration

# launch gateway service
mlflow gateway start --config-path examples/gateway/openai/config.yaml

# Then, run the examples in this PR
```

### CODE REVIEW

File: docs/extras/ecosystem/integrations/mlflow.mdx [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# MLflow Gateway
+
+## Installation and Setup
+
+Install `mlflow` with `gateway` extra:
+
+```sh
+pip install 'mlflow[gateway]'
```
Comment (harupy) [inline docs/extras/ecosystem/integrations/mlflow.mdx:8]: MLflow 2.5 is not released yet, we need to use the following command:

```
pip install git+https://github.com/mlflow/mlflow.git@gateway
```

File: langchain/llms/mlflow_gateway.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            **self.params.dict(exclude_none=True),
+        }
+        if s := (stop or self.params.stop):
+            data["stop"] = s
+        resp = mlflow.gateway.query(self.route, data=data)
+        return resp["candidates"][0]["text"]
+
+    @property
+    def _llm_type(self) -> str:
+        return "mlflow-gateway"
```
Comment (harupy) [inline langchain/llms/mlflow_gateway.py:59]: @dbczumar @BenWilson2

```suggestion
        return "mlflow-ai-gateway"
```

should we use `mlflow-ai-gateway`?

**Comment by baskaryan** [comment]
@harupy just ping me when it's ready for review

File: langchain/llms/mlflow_gateway.py (hunk 2 above)
Comment (BenWilson2) [reply langchain/llms/mlflow_gateway.py:59]: +1 to just match it to what people will see in docs investigations.

File: docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx [hunk 3]
```diff
@@ -0,0 +1,84 @@
+# MLflow AI Gateway
+
```
Comment (BenWilson2) [inline docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx:2]: Can we add a link to our (expected) docs page for the AI Gateway?

File: langchain/llms/mlflow_gateway.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+from langchain.callbacks.manager import CallbackManagerForLLMRun
+from langchain.llms.base import LLM
+from pydantic import BaseModel, Extra
+
+
+class Params(BaseModel, extra=Extra.allow):
+    temperature: float = 0.0
+    stop: List[str] | None = None
+    max_tokens: Optional[int] = None
+    candidate_count: int = 5
```
Comment (BenWilson2) [inline langchain/llms/mlflow_gateway.py:12]: Should we leave this as a default 1? We do have a provider that only supports the value of max '1' and it seems like defaulting to the cheaper of all options is probably something that users would appreciate :)

Comment (BenWilson2) [reply langchain/llms/mlflow_gateway.py:12]: We also only ever return the first candidate from the `_call()` below anyway

Comment (harupy) [reply langchain/llms/mlflow_gateway.py]: sounds good!

File: langchain/llms/mlflow_gateway.py (hunk 2 above)
Comment (harupy) [reply langchain/llms/mlflow_gateway.py]: made the change!

File: docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+embeddings = MlflowAIGatewayEmbeddings(
+    gateway_uri="http://127.0.0.1:5000",
+    route="embeddings",
+)
+
+print(embeddings.embed_query("hello"))
+print(embeddings.embed_documents(["hello"]))
+```
+
+## Databricks-managed MLflow AI Gateway
```
Comment (BenWilson2) [inline docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx:88]:
```suggestion
## Databricks MLflow AI Gateway
```

File: docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+To use Databricks-managed MLflow AI Gateway, set the gateway URI to `databricks`:
```
Comment (BenWilson2) [inline docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx:90]: (In private preview, please contact Databricks representative to enroll in the preview)

File: langchain/llms/mlflow_ai_gateway.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    candidate_count: int = 1
+
+
+class MlflowAIGateway(LLM):
+    route: str
+    gateway_uri: Optional[str] = None
+    params: Optional[Params] = None
+
+    def __init__(self, **kwargs: Any):
+        import mlflow.gateway
```
Comment (baskaryan) [inline langchain/llms/mlflow_ai_gateway.py:24]: can we put this in a try/excpet with a nice error message that includes instructions to install mlflow gateway

**Comment by baskaryan** [comment]
one small comment, otherwise lgtm! happy to merge whenever you think it's ready @harupy @BenWilson2

File: langchain/llms/mlflow_ai_gateway.py (hunk 7 above)
Comment (harupy) [reply langchain/llms/mlflow_ai_gateway.py:24]: Great idea, I'll make that change!

**Comment by harupy** [comment]
@baskaryan Is there a way to preview the documentation?

**Comment by baskaryan** [comment]
> @baskaryan Is there a way to preview the documentation? I'd like to check what the doc looks like.

kicked off a preview build, should be visible in a few minutes

**Comment by harupy** [comment]
@baskaryan Thanks!

@BenWilson2 @dbczumar We can preview the mlflow ai gateway page here:

https://langchain-bz56yl7oe-langchain.vercel.app/docs/ecosystem/integrations/mlflow_ai_gateway

File: docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx [hunk 8]
```diff
@@ -0,0 +1,109 @@
+# MLflow AI Gateway
```
Comment (dbczumar) [inline docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx:1]: Can we also demonstrate how users can log these models using LangChain?

File: langchain/embeddings/mlflow_gateway.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
+from __future__ import annotations
+
+from typing import Any, Iterator, List, Optional
+
+from pydantic import BaseModel
+
+from langchain.embeddings.base import Embeddings
+
+
+def chunk(texts: List[str], size: int) -> Iterator[List[str]]:
```
Comment (dbczumar) [inline langchain/embeddings/mlflow_gateway.py:10]: should this be internal?

```suggestion
def _chunk(texts: List[str], size: int) -> Iterator[List[str]]:
```

**Review by dbczumar — APPROVED** [review]
LGTM with 2 nits! Thanks @harupy !

File: docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+print(embeddings.embed_query("hello"))
+print(embeddings.embed_documents(["hello"]))
+```
+
+## Databricks MLflow AI Gateway
+
+Databricks MLflow AI Gateway is in private preview.
+Please contact Databricks representative to enroll in the preview.
```
Comment (BenWilson2) [inline docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx:89]:
```suggestion
Please contact a Databricks representative to enroll in the preview.
```

**Review by BenWilson2 — APPROVED** [review]
LGTM!

File: langchain/embeddings/mlflow_gateway.py (hunk 9 above)
Comment (harupy) [reply langchain/embeddings/mlflow_gateway.py]: Updated!

File: docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx (hunk 8 above)
Comment (harupy) [reply docs/extras/ecosystem/integrations/mlflow_ai_gateway.mdx:1]: Updated the completions example!

**Pull request merged**